src/core/reference/include/openvino/reference/xattention.hpp

```diff
@@ -0,0 +1,349 @@
+                                    T* out,
+                                    const Shape& out_shape) {
+        OPENVINO_ASSERT(attention_scores_shape.size() == 3);  // [num_heads, query_antidiagonals, key_antidiagonals]
+        size_t antidiagonals_per_xattention_block = m_block_size / m_stride;
+        OPENVINO_ASSERT(attention_scores_shape[1] % antidiagonals_per_xattention_block == 0);
+        OPENVINO_ASSERT(attention_scores_shape[2] % antidiagonals_per_xattention_block == 0);
+
+        OPENVINO_ASSERT(out_shape[0] == attention_scores_shape[0]);
+        OPENVINO_ASSERT(out_shape[1] ==
+                        attention_scores_shape[1] / antidiagonals_per_xattention_block);  // query length, blocked
+        OPENVINO_ASSERT(out_shape[2] ==
+                        attention_scores_shape[2] / antidiagonals_per_xattention_block);  // key length, blocked
+
+        std::memset(out, 0, out_shape[0] * out_shape[1] * out_shape[2] * sizeof(T));
+
+        for (size_t head_idx = 0; head_idx < attention_scores_shape[0]; head_idx++) {
+            size_t in_head_offset = head_idx * attention_scores_shape[1] * attention_scores_shape[2];
+            size_t out_head_offset = head_idx * out_shape[1] * out_shape[2];
+            for (size_t query_len_idx = 0; query_len_idx < attention_scores_shape[1]; query_len_idx++) {
+                for (size_t key_len_idx = 0; key_len_idx < attention_scores_shape[2]; key_len_idx++) {
+                    size_t query_block_idx = query_len_idx / antidiagonals_per_xattention_block;
+                    size_t key_block_idx = key_len_idx / antidiagonals_per_xattention_block;
+                    auto target_block_sum_ptr = out + out_head_offset + query_block_idx * out_shape[2] + key_block_idx;
+                    *target_block_sum_ptr += *(attention_scores_data + in_head_offset +
+                                               query_len_idx * attention_scores_shape[2] + key_len_idx);
+                }
+            }
+        }
+    }
+
+    /** Selects the elements of the input tensor along the last two dimensions, indepently along the first dimension, so
+     * that the elements constitute a smallest subset constituting a sum portion no less than `threshold` of the total
+     * element sum.
+     * @param blocked_scores_data Pointer to the blocked score input.
+     * @param blocked_attention_scores_shape Shape of the blocked score input tensor. Expected shape is [num_heads,
+     * num_query_tokens / block_size, num_key_tokens / block_size]
+     * @return A vector of size `num_heads` of sets, each set containing pairs of block indices (.first is the block
+     * index along the query dimension, .second - along the key). Each set is the head-specific subset of blocks
+     * corresponding to the property described above.
+     */
+    XAttentionRetainedBlockIndicesForAllHeads get_block_indices_to_keep(const T* blocked_attention_scores_data,
```

Comment:
To make XAttention compatible with attention score-based eviction methods, we should enforce keeping last_query_size = 100 tokens, and calculate the attention mass excluding the last Q blocks to correctly apply the threshold for retained cumulative sum.

Reply:
I thought we disabled the xattention for a prompt chunk on the openvino.genai level if the chunk overlaps with the last dense token region:
https://github.com/openvinotoolkit/openvino.genai/blob/9ddbb07cfa48af645664280d095a1840e5c5bdcf/src/cpp/src/continuous_batching/scheduler.hpp#L613
In the PA op there is no information on what position each chunk occupies in the prompt, we'd have to set up an extra per-sequence input to the PA op just to tell xattention how many "last" tokens there are in the current chunk.